src/backend/session/mod.rs

```diff
@@ -0,0 +1,155 @@
+//!
+//! Abstraction of different session apis.
+//!
+//! Sessions provide a way for multiple graphical systems to run in parallel by providing
+//! mechanisms to switch between and handle device access and permissions for every running
+//! instance.
+//!
+//! They are crutial to allow unpriviledged processes to use graphical or input devices.
+//!
+//! The following mechanisms are currently provided:
+//!     - direct - legacy tty / virtual terminal kernel api
+//!
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::os::unix::io::RawFd;
+use nix::fcntl::OFlag;
+use wayland_server::StateProxy;
+
+/// General session interface.
+///
+/// Provides a way to open and close devices and change the active vt.
+pub trait Session {
+    /// Error type of the implementation
+    type Error: AsErrno;
+
+    /// Opens a device at the given `path` with the given flags.
+    ///
+    /// Returns a raw file descriptor
+    fn open(&mut self, path: &Path, flags: OFlag) -> Result<RawFd, Self::Error>;
+    /// Close a previously opened file descriptor
+    fn close(&mut self, fd: RawFd) -> Result<(), Self::Error>;
+
+    /// Change the currently active virtual terminal
+    fn change_vt(&mut self, vt: i32) -> Result<(), Self::Error>;
+
+    /// Check if this session is currently active
+    fn is_active(&self) -> bool;
+    /// Which seat this session is on
+    fn seat(&self) -> String;
+}
+
+/// Interface for registering for notifications for a given session.
+///
+/// Part of the session api which allows to get notified, when the given session
+/// gets paused or becomes active again. Any object implementing the `SessionObserver` trait
+/// may be registered.
+pub trait SessionNotifier {
+    /// Registers a given `SessionObserver`.
+    ///
+    /// Returns an id of the inserted observer, can be used to remove it again.
+    fn register<S: SessionObserver + 'static>(&mut self, signal: S) -> usize;
+    /// Removes an observer by its given id from `SessionNotifier::register`.
+    fn unregister(&mut self, signal: usize);
+
+    /// Check if this session is currently active
+    fn is_active(&self) -> bool;
+    /// Which seat this session is on
+    fn seat(&self) -> &str;
```

Comment:
Is there a reason why `Session::seat()` returns `String` but `SessionNotifier::seat()` returns `&str`?

Reply:
Yes, because of the implementations of `Rc<RefCell<Session>>` and `Arc<Mutex<Session>>`. It is just way easier to clone the String in that case. Do you have a better idea?